Add keyword search to filter example queries

IoannisNezis opened the pull request:
This PR adds a keyword-search to filter the example querys (by name).

| before | after |
| ----------|---------|
| ![20240306_18h35m47s_grim](https://github.com/ad-freiburg/qlever-ui/assets/74186504/1cdb991a-d535-4fe1-826d-143329342761) | ![20240306_17h53m32s_grim](https://github.com/ad-freiburg/qlever-ui/assets/74186504/567a3430-3bc0-4b81-b868-392dd5a8bc21) |

Each whitespace sperated word needs to appear in the result.

The search matches the keyword also `infix`
![20240306_18h12m52s_grim](https://github.com/ad-freiburg/qlever-ui/assets/74186504/206e60a6-447a-42a2-bc16-e15dd584b0d1)
 
When no result matches the query, a message is displayed.
![20240306_18h41m10s_grim](https://github.com/ad-freiburg/qlever-ui/assets/74186504/7b3ecc25-f145-4164-8f55-f61ca487d597)

IoannisNezis commented:
Maybe debouncing the search would be nice

IoannisNezis commented:
i also played around a bit with different highlight options:

hannahbast commented:
@IoannisNezis This is awesome! It's live on https://qlever.cs.uni-freiburg.de for some time now and I use it for almost every example query. The underlining looks great. A minor suggestion for further improvement: When one clicks on the `Example` button, the focus should be in the search field so that one can start typing right away without first clicking into the search field

hannahbast commented on backend/static/js/keywordSearch.js:
```diff
@@ -0,0 +1,76 @@
+// this function removes the highlight
```
A minor comment: Our comment style is normal sentences (capitalized and with a full stop). I know that a lot of the qlever-ui does not follow this rule yet. That's becase it's old. But nor reason not to stick to this for all new code.

IoannisNezis replied:
Ah, good to know!
I generated some docstrings with chatGPT.
Do the new comments satisfy the comment style?

```
/**
 * Removes highlighting from the input string by replacing all occurrences of
 * <span> elements with the class "keyword-search-highlight" with their inner content.
 *
 * @param {string} input_str - The input string possibly containing highlighted spans.
 * @returns {string} - The input string with highlighting removed.
 */
```

hannahbast replied:
We use ordinary `// ...` comments with full sentences. When referring to variable or class names in a comment, we put it in `...`. See the code on https://github.com/ad-freiburg/qlever for many examples of this (it's also not 100% consistent there, but 90% of the comments follow this convention).

hannahbast commented on backend/static/js/keywordSearch.js:
```diff
@@ -0,0 +1,76 @@
+  let return_str = unHighlight(input_str.toLowerCase());
+  // find matching sections
+  let matching_sections = [];
+  for (word of words){
+    let startIndex = 0;
+    while ((index = return_str.indexOf(word, startIndex)) > -1){
+      matching_sections.push([index, index + word.length]);
+      startIndex = index + word.length;
+    }
+  }
```
With `input_str.match(regex)` this code would be simpler and more powerful (one could also type regexes then, which doesn't harm)

IoannisNezis replied:
Thanks for the suggestion.
I wanted to revisit regex for a while now and this was a great opportunity!

While i agree that this part of the code (finding matching sections in one `example name` ) got simpler,
I feel like the other part (filtering `example names` that match all given keywords) got more complex and harder to read.
But i think it works now, and searching with regex-keywords is pretty cool.

I'm not 100% sure i understood the suggestion perfectly.
Does the new version behave as you would have expected?

hannahbast commented on backend/static/js/keywordSearch.js:
```diff
@@ -0,0 +1,76 @@
+  }
+}
+
+// this is taken from here: https://www.geeksforgeeks.org/implement-search-box-with-debounce-in-javascript/
+// its a decorator that makes sure a function is only called once within the given delay
```
In which sense is this a "decorator"?

IoannisNezis replied:
With "decorator" i was referring to a design pattern of a "wrapper" function, a higher order function that consumes and returns a function.
This allows to add behavior *before* and *after* the consumed function.
In `python` these wrapper functions are called decorators.
Since the `debounce` funciton is a wrapper function that adds behaviour, without altering the behavior of the consumed function, i  thought its appropriate to call it a "decorator".

In the commit [enhancement: better docstrings for keyword search](https://github.com/ad-freiburg/qlever-ui/pull/83/commits/0175a3986c1699579ef29f3897724b8f90bf7163)  i changed the comment and dropped the mention of a decorator.

hannahbast commented on backend/templates/index.html:
```diff
@@ -201,7 +201,11 @@ <h4 style="margin-top: 0px;">QLever UI Shortcuts:</h4>
-                  <ul class="dropdown-menu">
+                  <ul id="example-list" class="dropdown-menu">
+                    <div style="padding: 10px 20px;">
+                      <input id="example-keyword-search" oninput="filterExamplesDebounced(event)" type="text" placeholder="search for keywords" style="width: 100%"> </input>
```
Would be great to have the focus automatically set on this whenever the Example button is pressed

IoannisNezis replied:
@hannahbast thanks for the suggestion, i implemented this and it feels very smooth and convenient.

hannahbast reviewed (changes requested):
Looks great, a few minor suggestions. Using `match(regex)` for the search is not a must, but maybe you like it.

hannahbast commented:
@IoannisNezis This looks great now.

Another small request, which occurred to me already when trying this out the first time and still occurs to me now every time I used the new feature:

When selecting a query using the "Example" button one has to:

1. Press the Example button
2. Start typing some word fragments (thanks to your latest change, a click in the search field is no longer necessary)
3. Click on the desired query
4. Execute the query

For Step 4, there is the Shortut Ctrl+Return, which I always use to execute a query (after you type the query, you can just continue typing in instead of having to switch to the mouse again just for a single click).

But with the workflow above, I have to click, type, click, click

It would be awesome if one could select the query with either TAB or arrow-up / arrow down and then select it via Return (and another Ctrl+Return would then execute it). Interestingly, when pressing TAB once, one already gets to "Add new example", and then Return takes you to the corresponding page. So maybe realizing this with TAB is relatively little additional work.

IoannisNezis commented:
@hannahbast
I actually already started working on this, initially this is a bit of a pain.
I did not look into the tabindex direction yet, thanks for the idea.

At this point it might be worth mentioning that i am re-implementing parts of this library: https://select2.org/ which i initially choose not to use.

If there is a good reason not to use this library i would be really happy (sunken cost fallacy is kicking in really badly).

IoannisNezis commented:
I implemented the keyboard control.
It should now be possible to *search*, *select* and *execute* an example query while only clicking once.

I decided against the tabulator option and implemented the selection with arrow keys.

# Further improvements
I want to implement 2 further improvements:

## Scroll
The list of possible examples is a bit long.  
i want to make the list scrollable, instead of showing everything.

## Exclusive Hover
Currently one can select a example with the mouse **and** the arrow keys.  
I want to change this so that if one hovers an example only this example is highlighted.
If the arrow key is used after a element is hovered with the mouse it should propagate from this element.

IoannisNezis commented:
@hannahbast 
In the last 2 commits I implemented the formatting changes we discussed (typo, line-length, ...).

A few things:

1. What is the procedure for closing threads in PR's?
Should only the reviewer close them or can the assignee also close threads if its reasonable?

2.  In a previous comment i suggested two further improvements.
I would like to address them in separate PR's to keep this one at a manageable size.

hannahbast reviewed (approved):
Finally finishing this review and merging this. Thanks for the extremely useful addition, Ioannis!
